File: Cpp/410-split-array-largest-sum/solution.cpp
class Solution {
    // Helper function to count the number of splits possible for a given maxsum
    int possiblesplit(vector<int>& arr, int maxsum){
        int subarraysum = 0, split = 1;
        for(int i : arr){
            if(subarraysum + i <= maxsum){
                subarraysum += i;
            }
            else{
                split += 1;
                subarraysum = i;
            }
        }
        return split;
    }

public:
    // Main function to split the array into 'k' subarrays such that the maximum sum of any subarray is minimized
    int splitArray(vector<int>& nums, int k) {
        // Initialize the search range for binary search
        int low = *max_element(nums.begin(), nums.end()); // Minimum possible maximum sum
        int high = accumulate(nums.begin(), nums.end(), 0); // Maximum possible maximum sum (sum of all elements)

        // Binary search to find the minimum maximum sum
        while(low <= high){
            int mid = (low + high) / 2;
            // If the current mid value allows for more splits than 'k', increase the minimum possible maximum sum
            if(possiblesplit(nums, mid) > k){
                low = mid + 1;
            }
            // If the current mid value allows for 'k' or fewer splits, decrease the maximum possible maximum sum
            else{
                high = mid - 1;
            }
        }
        // Return the lowest possible maximum sum
        return low;
    }
};

File: Py/410-split-array-largest-sum/solution.py
from typing import List

class Solution:
    def possiblesplit(self, arr: List[int], maxsum: int) -> int:
        subarraysum = 0  # Initialize the sum of the current subarray
        split = 1  # Initialize the number of splits to 1 (at least one subarray)

        for i in arr:
            if subarraysum + i <= maxsum:
                subarraysum += i  # Add the current element to the subarray sum
            else:
                split += 1  # Increase the number of splits
                subarraysum = i  # Start a new subarray with the current element

        return split

    def splitArray(self, nums: List[int], k: int) -> int:
        low = max(nums)  # Minimum possible maximum sum (initially set to the maximum element)
        high = sum(nums)  # Maximum possible maximum sum (sum of all elements)

        while low <= high:
            mid = (low + high) // 2  # Calculate the middle value

            # Check if the current mid value allows for more splits than 'k'
            if self.possiblesplit(nums, mid) > k:
                low = mid + 1  # Increase the minimum possible maximum sum
            else:
                high = mid - 1  # Decrease the maximum possible maximum sum

        # Return the lowest possible maximum sum that satisfies the condition
        return low
